### ISSUE TITLE
Different Loss Functions and How They Effect Knowledge Distillation

### ISSUE THREAD

**Issue by fredericowieser** [BrownianNotion/BitDistiller#41]
## 1. Description

Currently, Bit-distiller uses CAKLD loss to train the model. While the justification for this is discussed in the paper we want to explore the effect of other popular loss functions in knowledge distillation especially to see if we observe different behaviour for lower bit models. We will train for fewer epochs and start by experimenting on a 2 bit model so that we can compare the performance of the 2 bit from bit distiller. Temperature is also a hyperparameter we wish to experiment with. 

Currently, we are implementing CE + CAKLD
`loss_ce = criterion(student_outputs, labels)`
`loss_kd_fw = F.kl_div(F.log_softmax(student_outputs/T, dim=1),
                           F.softmax(teacher_outputs/T, dim=1),
                           reduction='batchmean') * (T * T)`
 `loss_kd_bw = F.kl_div(F.log_softmax(teacher_outputs/T, dim=1),
                           F.softmax(student_outputs/T, dim=1),
                           reduction='batchmean') * (T * T)`

 `loss = loss_kd_bw*(1-gamma) + (gamma) *loss_kd_fw + loss_ce` 
## 2. Hypothesis

Advantages of different loss functions:
CAKLD: Mode covering and seeking balanced by confidence of teacher - aims to only be mode seeking where teacher is confident. 
CE+CAKLD: standard classification loss whilst transferring the teacher loss - very popular method.
MMD: Measures the distance between distributions in a reproducing kernel Hilbert space. Encourages overall distribution alignment; can be used for feature matching. 
See here for a more detailed table: https://docs.google.com/spreadsheets/d/1Yw0bFqLAsaO00SctGdNZJWnAU7TGmyCiR-7btU0dkFE/edit?usp=sharing
## 3. Results

[Hugging Face page](https://huggingface.co/Niks898/TinyLlama_v1.1_2bit_int_ce_plus_cakld_20) 
T=20
Task | Accuracy | Std Error | Acc Norm | Acc Norm StdErr
-- | -- | -- | -- | --
arc_challenge | 0.2150 | 0.01201 | 0.2457 | 0.01258
arc_easy | 0.3998 | 0.01005 | 0.3801 | 0.00996
hellaswag | 0.3121 | 0.00462 | 0.3617 | 0.00480
piqa | 0.6012 | 0.01142 | 0.6028 | 0.01142
winogrande | 0.5178 | 0.01404 | — | —

T=5

Task | Accuracy | Std Error | Acc Norm | Acc Norm StdErr
-- | -- | -- | -- | --
arc_challenge | 0.2184 | 0.01207 | 0.2406 | 0.01249
arc_easy | 0.3616 | 0.00989 | 0.3611 | 0.00986
hellaswag | 0.3109 | 0.00462 | 0.3644 | 0.00480
piqa | 0.6029 | 0.01146 | 0.6023 | 0.01141
winogrande | 0.5280 | 0.01403 | 0.6229 | 0.01403

T=2

Task | Accuracy | Std Error | Acc Norm | Acc Norm StdErr
-- | -- | -- | -- | --
arc_challenge | 0.2056 | 0.01181 | 0.2372 | 0.01243
arc_easy | 0.3451 | 0.00976 | 0.3455 | 0.00976
hellaswag | 0.3137 | 0.00463 | 0.3643 | 0.00480
piqa | 0.6137 | 0.01136 | 0.6028 | 0.01142
winogrande | 0.5138 | 0.01405 | — | —

## References

**Commit Hash:** 

**Reviewer:**
